src/core/gui/Layout.cpp

```diff
@@ -254,57 +285,163 @@ void Layout::layoutPages(int width, int height) {
+
+    // Iterate over ALL possible rows and columns 3 times:
+    // We don't know which page, if any, is to be displayed in each row, column -  ask the mapper object!
+    // Once we know the dimensions of all pages, define the paddings above and left of each page as required.
+    // Finally, assign page coordinates by propagating page paddings and sizes from the border.
+
+    // first loop: set mapped position in view object and retrieve page dimensions
     for (size_t r = 0; r < rows; r++) {
         for (size_t c = 0; c < columns; c++) {
-            auto optionalPage = this->mapper.at({c, r});
+            const auto index = c * rows + r;
+            pages[index].optionalPage = this->mapper.at({c, r});
 
-            if (optionalPage) {
+            if (pages[index].optionalPage) {
+                auto& v = this->view->viewPages[*pages[index].optionalPage];
 
-                auto& v = this->view->viewPages[*optionalPage];
                 v->setMappedRowCol(strict_cast<int>(r),
                                    strict_cast<int>(c));  // store row and column for e.g. proper arrow key navigation
-                auto vDisplayWidth = v->getDisplayWidthDouble();
-                {
-                    auto paddingLeft = 0.0;
-                    auto paddingRight = 0.0;
-                    auto columnPadding = this->pc.widthCols[c] - vDisplayWidth;
+
+                pages[index].width = v->getDisplayWidthDouble();
+                pages[index].height = v->getDisplayHeightDouble();
+            }
+        }
+    }
+
+    // second loop: set paddings, width and height according to the layout style and orientation
+    if (this->layoutType == LAYOUT_TYPE_GRID) {
+        for (size_t r = 0; r < rows; r++) {
+            for (size_t c = 0; c < columns; c++) {
+                const auto index = c * rows + r;
+
+                auto columnPadding = this->pc.widthCols[c] - pages[index].width;
+
+                if (isPairedPages && len > 1) {
+                    // pair pages mode
+                    if (c % 2 == 0) {
+                        // align right
+                        pages[index].paddingLeft = XOURNAL_PADDING_BETWEEN - XOURNAL_ROOM_FOR_SHADOW + columnPadding;
+                    } else {  // align left
+                        pages[index].paddingLeft = XOURNAL_ROOM_FOR_SHADOW;
+                    }
+                } else {  // not paired page mode - center
+                    pages[index].paddingLeft = XOURNAL_PADDING_BETWEEN / 2.0 + columnPadding / 2.0;  // center justify
+                }
+
+                // center page vertically
+                pages[index].paddingTop = (this->pc.heightRows[r] - pages[index].height) / 2.0;
+
+                pages[index].width = pc.widthCols[c];
+                pages[index].height = pc.heightRows[r];
+            }
+        }
+    } else if (this->layoutType == LAYOUT_TYPE_CONST_PADDING) {
```

Comment:
rather `else if (...)` or a simple `else` like on line 428? In any case it should be consistent.